qiskit_experiments/framework/backend_timing.py

```diff
@@ -0,0 +1,349 @@
+        """
+        backend_data = BackendData(backend)
+
+        # Pull all the timing data from the backend
+        self._backend_data_dt = backend_data.dt
+        self._acquire_alignment = backend_data.acquire_alignment
+        self._granularity = backend_data.granularity
+        self._min_length = backend_data.min_length
+        self._pulse_alignment = backend_data.pulse_alignment
+
+    @property
+    def delay_unit(self) -> str:
+        """The delay unit for the current backend
+
+        "dt" is used if dt is present in the backend configuration. Otherwise
+        "s" is used.
+        """
+        if self._backend_data_dt is not None:
+            return "dt"
+
+        return "s"
+
+    @property
+    def _dt(self) -> float:
+        """Backend dt value
+
+        This property wraps :code:`.BackendData.dt` in order to give a more
+        specific error message when trying to use ``dt`` with a backend that
+        does not provide it rather than just giving a ``TypeError`` about
+        ``NoneType``. As this raises an exception when ``dt`` is not set, it
+        likely should not be used by external code using
+        :class:`.BackendTiming`.
+
+        Raises:
+            QiskitError: The backend does not include a dt value.
+        """
+        if self._backend_data_dt is not None:
+            return float(self._backend_data_dt)
+
+        raise QiskitError("Backend has no dt value.")
+
+    def circuit_delay(self, time: float) -> Union[int, float]:
+        """Delay duration close to ``time`` and consistent with timing constraints
+
+        This method produces the value to pass for the ``duration`` of a
+        ``Delay`` instruction of a ``QuantumCircuit`` schedule so that the
+        delay fills the time until the next valid pulse, assuming the ``Delay``
+        instruction begins on a sample that is also valid for a pulse to begin
+        on.
+
+        The pulse timing constraints of the backend are considered in order to
+        give a number of samples closest to ``time`` plus however many more
+        samples are needed to get to the next valid sample for the start of a
+        pulse in a subsequent instruction. The least common multiple of the
+        pulse and acquire alignment values is used in order to ensure that
+        either type of pulse will be aligned.
+
+        If :meth:`.BackendTiming.delay_unit` is ``s``, ``time`` is
+        returned directly. Typically, this is the case for a simulator where
+        converting to sample number is not needed.
+
+        Args:
+            time: The nominal delay time to convert in seconds
+
+        Returns:
+            The delay duration in samples if :meth:`.BackendTiming.delay_unit`
+            is ``dt``. Other return ``time``.
+        """
+        if self.delay_unit == "s":
+            return time
+
+        return self.schedule_delay(time)
+
+    def schedule_delay(self, time: float) -> int:
```

Comment:
Does this need to be a public? To me this looks like a protected member.

Reply:
My thinking was that you would use this method if calculating the delay within a schedule and the other for the delay in a circuit. The only difference is that the circuit one can give back the time in seconds if the backend does not support dt. That behavior doesn't make since for a pulse schedule, so I separated the two use cases.